url_shortener_proj/app/Http/Controllers/UrlController.php

```diff
@@ -0,0 +1,55 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use App\Url;
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\DB;
+use Str;
+use App\Visitor;
+use Hash;
+use Session;
+
+
+class UrlController extends Controller
+{
+    public function store(Request $request)
+    {
+        $request->validate([
+            'original' => 'required|url'
+        ]);
+        $url = Url::create([
+            'original' => $request->original,
+            'short' => Str::random(7)
+        ]);
+        $short = $url->short;
+
+        Session::flash('success', 'The short URL was generated successfully!');
+
+        return redirect()->route('urls.details', $short);
+    }
+
+    public function details($short)
+    {
+        $urls = Url::all();
+        foreach ($urls as $u) {
+            if ($u->short == $short)
+                $url = $u;
+        }
```

Comment:
This approach is not very efficient. You are loading all rows from urls table (`select * from urls`), Laravel will create `Url` objects for every loaded row and put all those objects into collection.

After that, you iterate over collection and compare "short" property using PHP. If you have tens of thousands of rows, this will be slow and will consume lots of memory.

Can you use the database to fetch only one row?